kong/clustering/services/sync/rpc.lua

```diff
@@ -0,0 +1,302 @@
+    if premature then
+      return
+    end
+
+    local res, err = concurrency.with_worker_mutex(SYNC_MUTEX_OPTS, function()
+      -- here must be 2 times
+      for i = 1, 2 do
+        local ns_deltas, err = kong.rpc:call("control_plane", "kong.sync.v2.get_delta",
+                                             { default =
+                                               { version =
+                                                 tonumber(declarative.get_current_hash()) or 0,
+                                               },
+                                             })
+        if not ns_deltas then
+          ngx_log(ngx_ERR, "sync get_delta error: ", err)
+          return true
+        end
+
+        local version = 0
+
+        local crud_events = {}
+        local crud_events_n = 0
+
+        local ns_delta
+
+        for namespace, delta in pairs(ns_deltas) do
+          if namespace == "default" then
+            ns_delta = delta
+          end
+        end
+
+        if not ns_delta then
+          return nil, "default namespace does not exist inside params"
+        end
+
+        if #ns_delta.deltas == 0 then
+          ngx_log(ngx_DEBUG, "no delta to sync")
+          return true
+        end
+
+        local t = txn.begin(512)
+
+        if ns_delta.wipe then
+          t:db_drop(false)
+        end
+
+        for _, delta in ipairs(ns_delta.deltas) do
+          if delta.row ~= ngx.null then
+            -- upsert the entity
+            -- does the entity already exists?
+            local old_entity, err = kong.db[delta.type]:select(delta.row)
+            if err then
+              return nil, err
+            end
+
+            local crud_event_type = "create"
+
+            if old_entity then
+              local res, err = delete_entity_for_txn(t, delta.type, old_entity, nil)
+              if not res then
+                return nil, err
+              end
+
+              crud_event_type = "update"
+            end
+
+            local res, err = insert_entity_for_txn(t, delta.type, delta.row, nil)
+            if not res then
+              return nil, err
+            end
+
+            crud_events_n = crud_events_n + 1
+            crud_events[crud_events_n] = { delta.type, crud_event_type, delta.row, old_entity, }
+
+          else
+            -- delete the entity
+            local old_entity, err = kong.db[delta.type]:select({ id = delta.id, }) -- TODO: composite key
+            if err then
+              return nil, err
+            end
+
+            if old_entity then
+              local res, err = delete_entity_for_txn(t, delta.type, old_entity, nil)
+              if not res then
+                return nil, err
+              end
+            end
+
+            crud_events_n = crud_events_n + 1
+            crud_events[crud_events_n] = { delta.type, "delete", old_entity, }
+          end
+
+          -- XXX TODO: could delta.version be nil or ngx.null
```

Comment:
This doesn't sounds right, why would `delta.version` be null?

Reply:
Got, we will investigate this later.

Reply:
KAG-5493

Reply:
resolved this, we could fix it in later PR, and has a kag to track this